pkg/maintainer/wallet/wallet_test.go

```diff
@@ -0,0 +1,72 @@
+package wallet
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/keep-network/keep-core/pkg/tbtc"
+)
+
+func TestRunIfWalletUnlocked_WhenLocked(t *testing.T) {
+	localChain := newLocalChain()
+
+	walletPublicKeyHash := [20]byte{1}
+
+	lockExpiration := time.Now().Add(500 * time.Millisecond)
+
+	localChain.setWalletLock(
+		walletPublicKeyHash,
+		lockExpiration,
+		tbtc.Heartbeat,
+	)
+
+	runFunc := func() error {
+		return fmt.Errorf("boom, you should not run me")
+	}
+
+	wm := &walletMaintainer{
+		config:   Config{},
+		chain:    localChain,
+		btcChain: newLocalBitcoinChain(),
+	}
+
+	err := wm.runIfWalletUnlocked(
+		context.Background(),
+		walletPublicKeyHash,
+		tbtc.DepositSweep,
+		runFunc,
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestRunIfWalletUnlocked_WhenUnlocked(t *testing.T) {
+	localChain := newLocalChain()
+
+	walletPublicKeyHash := [20]byte{2}
+
+	localChain.resetWalletLock(walletPublicKeyHash)
+
+	runFunc := func() error {
+		return nil
+	}
+
+	wm := &walletMaintainer{
+		config:   Config{},
+		chain:    localChain,
+		btcChain: newLocalBitcoinChain(),
+	}
+
+	err := wm.runIfWalletUnlocked(
+		context.Background(),
+		walletPublicKeyHash,
+		tbtc.DepositSweep,
+		runFunc,
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
```

Comment:
Note that both tests in this file check only if `err != nil` and they don't really check if the function was called.

I propose we add one more test covering the error propagation and add a condition to `_WhenUnlocked` test checking if the `runFunc` was actually called.

```
func TestRunIfWalletUnlocked_WhenUnlocked(t *testing.T) {
	localChain := newLocalChain()

	walletPublicKeyHash := [20]byte{2}

	localChain.resetWalletLock(walletPublicKeyHash)

	wasCalled := new(atomic.Bool)
	runFunc := func() error {
		wasCalled.Swap(true)
		return nil
	}

	wm := &walletMaintainer{
		config:   Config{},
		chain:    localChain,
		btcChain: nil,
	}

	err := wm.runIfWalletUnlocked(
		context.Background(),
		walletPublicKeyHash,
		tbtc.DepositSweep,
		runFunc,
	)
	if err != nil {
		t.Fatal(err)
	}
	if !wasCalled.Load() {
		t.Fatal("the function was not called")
	}
}

func TestRunIfWalletUnlocked_WhenUnlocked_PropagateError(t *testing.T) {
	localChain := newLocalChain()

	walletPublicKeyHash := [20]byte{2}

	localChain.resetWalletLock(walletPublicKeyHash)

	expectedError := fmt.Errorf("this was expected")
	runFunc := func() error {
		return expectedError
	}

	wm := &walletMaintainer{
		config:   Config{},
		chain:    localChain,
		btcChain: nil,
	}

	err := wm.runIfWalletUnlocked(
		context.Background(),
		walletPublicKeyHash,
		tbtc.DepositSweep,
		runFunc,
	)
	testutils.AssertErrorsSame(t, expectedError, err)
}
```

Reply:
Good catch! Unfortunately, atomic Bool was not available in Go 1.18, so a decided to sync via a channel. fe2f943588d1d57e9063feffa709994d005a99e6